fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryManager.java

```diff
@@ -73,6 +87,70 @@ public synchronized void close() throws IOException {
         createdDirectories.clear();
     }
 
+    @SuppressWarnings("PMD.CloseResource")
+    public CompletableFuture<Void> mergeIndex(LuceneAnalyzerWrapper analyzerWrapper) {
+
+        final ScanProperties scanProperties = ScanProperties.FORWARD_SCAN.with(
+                props -> props.clearState().setReturnedRowLimit(1));
+        final Range range = state.indexSubspace.range();
+        final KeyRange keyRange = new KeyRange(range.begin, range.end);
+
+        final RecordCursor<Tuple> cursor = new ChainedCursor<>(
+                state.context,
+                lastKey -> nextTuple(state.context, state.indexSubspace, keyRange, lastKey, scanProperties),
+                Tuple::pack,
+                Tuple::fromBytes,
+                null,
+                scanProperties);
+
+        return cursor.forEach(groupingKey -> mergeIndex(analyzerWrapper, groupingKey));
+    }
+
+    private void mergeIndex(LuceneAnalyzerWrapper analyzerWrapper, Tuple groupingKey) {
+        try {
+            getDirectoryWrapper(groupingKey).mergeIndex(analyzerWrapper);
+        } catch (IOException e) {
+            if (LOG.isWarnEnabled()) {
+                LOG.warn("Lucene mergeIndex failed: error={}", e.getMessage());
+            }
+        }
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
+    private CompletableFuture<Optional<Tuple>> nextTuple(@Nonnull FDBRecordContext context,
+                                                         @Nonnull Subspace subspace,
+                                                         @Nonnull KeyRange range,
+                                                         @Nonnull Optional<Tuple> lastTuple,
+                                                         @Nonnull ScanProperties scanProperties) {
+        final KeyValueCursor cursor =
+                lastTuple.isEmpty()
+                ?
+                KeyValueCursor.Builder.withSubspace(subspace)
+                        .setContext(context)
+                        .setRange(range)
+                        .setContinuation(null)
+                        .setScanProperties(scanProperties)
+                        .build()
+                :
+                KeyValueCursor.Builder.withSubspace(subspace)
+                        .setContext(context)
+                        .setContinuation(null)
+                        .setScanProperties(scanProperties)
+                        .setLow(subspace.pack(lastTuple.get().get(0)), EndpointType.RANGE_EXCLUSIVE)
```

Comment:
I don't think this will work correctly when there is more than one grouping key. You'll have to get the size of the grouping key from the root expression of the index, and set that sub-tuple as the low here.

Reply:
✅